// virtualized/source/utils/browser.js
/**
 * Browser Utils and APIs
 * For more information see https://plainjs.com/javascript/styles/getting-width-and-height-of-an-element-23/
 * @flow
 */

/**
 * https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle
 * Used in order for browser dependencies to be easily stubbed in tests
 * We wrap the globals to prevent "Illegal invocation" errors
 */
export function getComputedStyle (...args) {
  return global.getComputedStyle(...args)
}

/**
 * https://developer.mozilla.org/en-US/docs/Web/API/Element
 * Used in order for browser dependencies to be easily stubbed in tests
 * We wrap the globals to prevent "Illegal invocation" errors
 */
export const Element = global.Element

/** Returns the root document element. */
export function getDocumentRoot (): Element {
  return document.documentElement || document.body
}

/** Get window width */
export function getWindowWidth (): number {
  return window.innerWidth ||
    getDocumentRoot().clientWidth
}

/** Get window height */
export function getWindowHeight (): number {
  return window.innerHeight ||
    getDocumentRoot().clientHeight
}

/** Get element padding */
export function getPadding (element: Element): { [key: string]: number } {
  const {
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft
  } = getComputedStyle(element)

  return {
    paddingTop: parseInt(paddingTop, 10) || 0,
    paddingRight: parseInt(paddingRight, 10) || 0,
    paddingBottom: parseInt(paddingBottom, 10) || 0,
    paddingLeft: parseInt(paddingLeft, 10) || 0
  }
}
