// virtualized/karma.conf.js
// Reference: http://karma-runner.github.io/0.12/config/configuration-file.html
module.exports = function karmaConfig (config) {
  config.set({

    frameworks: [
      // Reference: https://github.com/karma-runner/karma-mocha
      'mocha'
    ],

    reporters: [
      // Reference: https://github.com/mlex/karma-spec-reporter
      'spec',

      // Reference: https://github.com/karma-runner/karma-junit-reporter
      'junit'
    ],

    files: [
      // Load polyfills and all tests
      'test.config.js'
    ],

    preprocessors: {
      // Reference: http://webpack.github.io/docs/testing.html
      // Reference: https://github.com/webpack/karma-webpack
      'test.config.js': ['webpack', 'sourcemap']
    },

    browsers: [
      // Run tests using PhantomJS
      'PhantomJS'
    ],

    junitReporter: {
      outputDir: (process.env.CIRCLE_TEST_REPORTS || 'public') + '/karma',
      suite: 'karma'
    },

    singleRun: true,

    plugins: [
      require('karma-mocha'),
      require('karma-webpack'),
      require('karma-spec-reporter'),
      require('karma-junit-reporter'),
      require('karma-sourcemap-loader'),
      require('karma-phantomjs-launcher')
    ],

    webpack: require('./webpack.test'),

    webpackMiddleware: {
      noInfo: true
    }
  })
}
